### PR TITLE
Introducing apprunning message version 2

### PR DESCRIPTION
**Pull request by Cabecinha84**
- New fluxapprunning message version 2;
- Version 2 will be used when a fluxNode is running more than one app, instead of sending one message for each app to the network, it agregates all information in just one message (version2).

### CODE REVIEW

File: ZelBack/src/services/appsService.js [hunk 1]
```diff
@@ -6034,7 +6034,20 @@ async function storeAppRunningMessage(message) {
...
+
+  if (message.version !== 1 && message.version !== 2) {
+    return new Error(`Invalid Flux App Running message for storing version ${message.version} not supported`);
+  }
+
+  if (message.version === 1 && (typeof message.hash !== 'string' || typeof message.name !== 'string')) {
+    return new Error('Invalid Flux App Running message for storing');
+  }
+
+  if (message.version === 2 && (!message.apps || !Array.isArray(message.apps))) {
```
Comment (TheTrunk) [inline ZelBack/src/services/appsService.js:6050]: Here we should also check the typo of every item in the array

File: ZelBack/src/services/appsService.js [hunk 2]
```diff
@@ -6056,30 +6069,71 @@ async function storeAppRunningMessage(message) {
...
+    const options = {
+      upsert: true,
+    };
+    await dbHelper.updateOneInDatabase(database, globalAppsLocations, queryUpdate, update, options);
+  } else {
+    let messageNotOk = false;
+    for (let i = 0; i < message.apps.length; i += 1) {
+      const app = message.apps[i];
+      if (typeof app.hash !== 'string' || typeof app.name !== 'string') {
+        return new Error('Invalid Flux App Running v2 message for storing');
```
Comment (TheTrunk) [inline ZelBack/src/services/appsService.js:6102]: This type check is what should happen earlier (just so we have type validation at one place)

File: ZelBack/src/services/appsService.js [hunk 3]
```diff
@@ -6056,30 +6069,71 @@ async function storeAppRunningMessage(message) {
...
+      };
+
+      // indexes over name, hash, ip. Then name + ip and name + ip + broadcastedAt.
+      const queryFind = { name: newAppRunningMessage.name, ip: newAppRunningMessage.ip, broadcastedAt: { $gte: newAppRunningMessage.broadcastedAt } };
+      const projection = { _id: 0 };
+      // we already have the exact same data
+      // eslint-disable-next-line no-await-in-loop
+      const result = await dbHelper.findOneInDatabase(database, globalAppsLocations, queryFind, projection);
+      if (result) {
+        // found a message that was already stored/bad message
```
Comment (TheTrunk) [inline ZelBack/src/services/appsService.js:6119]: this was only stored, can't be a bad message I think

File: ZelBack/src/services/appsService.js [hunk 4]
```diff
@@ -6056,30 +6069,71 @@ async function storeAppRunningMessage(message) {
...
+        messageNotOk = true;
+        break;
```
Comment (TheTrunk) [inline ZelBack/src/services/appsService.js:6121]: Here we should do continue to process the other parts of the message, apps that can be running

File: ZelBack/src/services/appsService.js [hunk 5]
```diff
@@ -6056,30 +6069,71 @@ async function storeAppRunningMessage(message) {
...
+      const queryUpdate = { name: newAppRunningMessage.name, ip: newAppRunningMessage.ip };
+      const update = { $set: newAppRunningMessage };
+      const options = {
+        upsert: true,
+      };
+      // eslint-disable-next-line no-await-in-loop
+      await dbHelper.updateOneInDatabase(database, globalAppsLocations, queryUpdate, update, options);
+    }
+    if (messageNotOk) {
+      return false;
```
Comment (TheTrunk) [inline ZelBack/src/services/appsService.js:6132]: Correct, I agree that rebroadcast should happen only in the case when we did not find any of the running messages in our database. If we find a single one, we should not rebroadcast

File: ZelBack/src/services/appsService.js (hunk 4 above)
Comment (TheTrunk) [reply ZelBack/src/services/appsService.js:6121]: Actually OK. Lets just break it and not process further.

File: ZelBack/src/services/appsService.js (hunk 3 above)
Comment (TheTrunk) [reply ZelBack/src/services/appsService.js:6119]: can leave it

File: ZelBack/src/services/appsService.js [hunk 6]
```diff
@@ -8347,13 +8401,25 @@ async function checkAndNotifyPeersOfRunningApps() {
...
     });
-    // eslint-disable-next-line no-restricted-syntax
-    for (const application of installedAndRunning) {
-      log.info(`${application.name} is running properly. Broadcasting status.`);
-      try {
+
+    const syncStatus = daemonServiceMiscRpcs.isDaemonSynced();
+    const daemonHeight = syncStatus.data.height || 0;
+
+    if (daemonHeight >= config.fluxapps.apprunningv2 && installedAndRunning.length > 1) {
```
Comment (TheTrunk) [inline ZelBack/src/services/appsService.js:8408]: to be simplified to not have repetitive code

**Review by TheTrunk — CHANGES_REQUESTED** [review]
1. Processing logic of running messages OK. 
- Suggestion is to move the validation check to front so its all grouped together.
(I was thinking of disallowing other keys but its not needed as its not getting stored) 
2. Logic of the function checkAndNotifyPeersOfRunningApps is all OK. 
- The suggestion is to simplify the function as there is some repetitive code included in both if and else block.
3. Test suite - some more tests are broken with this and it would be great to cover the additional logic added

File: ZelBack/src/services/appsService.js (hunk 6 above)
Comment (Cabecinha84) [reply ZelBack/src/services/appsService.js:8396]: ohh, I have missed this one... let me see what can I do

**Comment by Cabecinha84** [comment]
@TheTrunk 
All things adjusted/changed. I thinks it's more cleaner now.
TestSuite don't have tests for the fluxapprunning.
Is one of the things I have on the list to improve is the TestSuite, for sure lots of tests can be added. Once I have a computer that can run tests locally.

**Review by TheTrunk — APPROVED** [review]
Nicely done :+1:

**Pull request merged**